skill_tagging/skill_tagging_mixin.py

```diff
@@ -0,0 +1,82 @@
+"""
+A mixin that fetches and verifies skills related to an Xblock, that can be added for all XBlocks.
+"""
+
+import logging
+from django.conf import settings
+from django.utils.translation import gettext as _
+from openedx_events.learning.data import XBlockSkillVerificationData
+from openedx_events.learning.signals import XBLOCK_SKILL_VERIFIED
+from urllib.parse import quote, urljoin
+from xblock.core import XBlock
+from xblock.fields import Boolean, Scope
+
+from .utils import get_api_client
+
+LOGGER = logging.getLogger(__name__)
+
+# Make '_' a no-op so we can scrape strings
+_ = lambda text: text
+
+
+@XBlock.wants('user')
+class SkillTaggingMixin:
+    has_verified_tags = Boolean(
+        display_name=_("Has verified tags"),
+        default=False,
+        help=_("Xblock has verified tags?"),
+        scope=Scope.user_state
+    )
+    
+    def _fetch_skill_tags(self):
+        if hasattr(settings, 'TAXONOMY_API_BASE_URL'):
+            API_BASE_URL = settings.TAXONOMY_API_BASE_URL
+        else:
+            API_BASE_URL = None
+
+        if not API_BASE_URL:
+            LOGGER.warning("Require TAXONOMY_API_BASE_URL to be present in the settings.")
+            return
+
+        user_service = self.runtime.service(self, 'user')
+        if not user_service:
+            LOGGER.info("No user service available for this xblock. Cannot proceed.")
+            return
+
+        user = user_service.get_current_user()
+        api_client = get_api_client(user=user)
+
+        usage_id_str = self.scope_ids.usage_id.__str__()
+        XBLOCK_SKILL_TAGS_API = urljoin(
+            API_BASE_URL,
+            '/taxonomy/api/v1/xblocks/?usage_key={}'.format(quote(usage_id_str))
+        )
+        response = api_client.get(XBLOCK_SKILL_TAGS_API)
+        response.raise_for_status()
+        result = response.json()['results'][0]
+        return result['skills']
```

Comment:
It is possible that an xblock doesn't have any tags, we would need to check.

Reply:
@navinkarkera In that case `result['skills']` will just be an empty list wouldn't it? Does it need to log any error message if it's empty?

Reply:
@pkulkark I think we don't have `result` inside response json at all. `response.json()` will be a list with data like below
```json
[
  {
    "id": 1,
    "skills": [
      {
        "id": 1,
        "name": "SKILL-0"
      },
      {
        "id": 2,
        "name": "SKILL-1"
      },
      {
        "id": 3,
        "name": "SKILL-2"
      },
      {
        "id": 4,
        "name": "SKILL-3"
      },
      {
        "id": 5,
        "name": "SKILL-4"
      }
    ],
    "usage_key": "prove-home-street-v1:edx+prove-home-street+prove-home-street+prove-home-street@prove-home-street",
    "requires_verification": true,
    "auto_processed": false,
    "hash_content": null
  }
]
```

>In that case result['skills'] will just be an empty list wouldn't it? Does it need to log any error message if it's empty?

It is possible that `response.json()` returns an empty list so `response.json()[0]` will fail.

```suggestion
        result = response.json()
        if not result:
            return []
        return result[0].get('skills', [])
```

Reply:
@navinkarkera I implemented this based on the output I saw on my devstack for the api `/taxonomy/api/v1/xblocks/?usage_key=<xblock-usage-key>`. The output does contain `result` object:

```
{
  "count": 1,
  "next": null,
  "previous": null,
  "results": [
    {
      "id": 1,
      "skills": [
        {
          "id": 1,
          "name": "SKILL-0"
        },
        {
          "id": 6,
          "name": "SKILL-5"
        },
        {
          "id": 11,
          "name": "SKILL-10"
        },
        {
          "id": 16,
          "name": "SKILL-15"
        }
      ],
      "usage_key": "how-brother-accept-v1:edx+how-brother-accept+how-brother-accept+how-brother-accept@how-brother-accept",
      "requires_verification": true,
      "auto_processed": false,
      "hash_content": null
    }
  ]
}
```
I assume the `result` object will just contain an empty list if the xblock has no skills? In that case, the `response.json()['results'][0]` might still throw an error, so i'll add a check for it.

I think what you are referring to is for the api `/taxonomy/api/v1/xblocks/{id}/`, which requires the id of the `XBlockSkill` object. I went with the first api option since the usage key is more easily available for the mixin, than the related `XBlockSkill` id. Please let me know if you have any objections with using that one.

Reply:
>I implemented this based on the output I saw on my devstack for the api /taxonomy/api/v1/xblocks/?usage_key=<xblock-usage-key>

@pkulkark Ahh, my bad, I was referring to these [api test cases](https://github.com/openedx/taxonomy-connector/blob/2c7a2b1fa1198fd09be39e57f2db79edeac30a6f/tests/test_views.py#L509-L533).
If we are receiving paginated response, shouldn't we loop and get all the pages i.e. all the skills?

Reply:
@navinkarkera Ah you're right. I forgot about that. I've updated it to use the api as shown in the api test cases you linked. And removed the `result` object. Does it look ok now?

Reply:
@pkulkark No, no. I meant that I was wrong. In tests the results are not paginated hence I was confused.

I made some changes after testing in devstack but I am unable to push to this repo due to permission issue, hence created this MR: https://github.com/open-craft/xblock-skill-tagging/pull/2
Please merge it and then we can merge this MR.

Reply:
@navinkarkera Ah sorry I misunderstood your comment. And thank you for the fix. I've merged the PR now.